### ISSUE TITLE
Add runtime validation to models 

### ISSUE THREAD

**Issue by 2color** [prisma/prisma#3528]
## Problem
Before saving data to the database, it's common to perform validation. 

When using Prisma to save user data (in an API), Prisma will perform runtime validation before saving to the database and throw an error if the user data doesn't match the type. 

Even with TypeScript, these checks cannot happen at compile time because the user data is not known. 

Since there are situations where you want to validate the data as a separate step to saving to the database, it'd be nice if Prisma exposed the same validation functionality as a standalone method on models.

## Suggested solution
Provide a validate function for models which is based on the Prisma schema:

```prisma
model User {
  id        Int     @default(autoincrement()) @id
  email     String  @unique
  firstName String?
  lastName  String?
  social    Json?
  isAdmin   Boolean @default(false)
}
```

API ideas:
```ts
const { isValid, error } = prisma.user.validate(user)
```

The idea is that If the input is valid, then the error will be undefined. If the input is invalid, `error` is assigned an Error object providing more information.

OR 

```ts
const user = { firstName: 'Laura' }
try {
  const validatedUser = await prisma.user.validate(user)
} catch (error) {
  // handle error
}
```

## Additional context

For inspiration, [Joi's validation errors](https://joi.dev/api/?v=17.2.1#validationerror) might be useful (albeit more extensive as Joi has richer validation rules).

**Comment by barczaG** [comment]
We would need the ability to add custom validation rules.
Example use case: Storing field like social security number where you have custom business logic to validate if the number is valid (eg. the sum of first and fifth digit must be equal sixth digit modulo 10)

**Comment by orenmizr** [comment]
so.... until that happens.... what is the best way to do model validation today with prisma 2 ?

**Comment by 2color** [comment]
> so.... until that happens.... what is the best way to do model validation today with prisma 2 ?

Hey @orenmizr,

Prisma validates input at runtime and throws an exception if validation fails (and the input data doesn't match the type in the Prisma schema)

This is very useful in both JavaScript projects where there are no compile-time checks and TypeScript projects where payload/input data cannot be type-checked at compile time. 

## Example

Given the following Prisma schema: 

https://github.com/prisma/prisma-examples/blob/b5c3aaa8df27c39422ea64525f513e025c8358c7/javascript/script/prisma/schema.prisma#L10-L24

This is the error that will be thrown with mistyped input:

```
PrismaClientValidationError:
Invalid `prisma.user.create()` invocation:

{
  data: {
    email: 'alice@prisma.io',
    name: 1,
          ~
    posts: {
      create: {
        title: 'Watch the talks from Prisma Day 2019',
        content: 'https://www.prisma.io/blog/z11sg6ipb3i1/',
        published: 'hello'
                   ~~~~~~~
      }
    }
  },
  include: {
    posts: true
  }
}

Argument name: Got invalid value 1 on prisma.createOneUser. Provided Int, expected String or Null.
Argument published: Got invalid value 'hello' on prisma.createOneUser. Provided String, expected Boolean.
```

**Comment by orenmizr** [comment]
thanks for replying 🙏
not sure if that is the original intent of the author - but i am looking for some validation into the model like in typeORM. you get an object over the network and you want to know whether you should continue with DB update.

```
@Column()
    @IsEmail()
    email: string;

    @Column()
    @IsDate()
    createDate: Date;
```

**Comment by voycey** [comment]
I'm unsure why this is not a thing - every good ORM should have the ability for easy validation, are there any 3rd party libraries that interface cleanly with Prisma for this?

**Comment by mmahalwy** [comment]
I guess another workaround is using something like `yup` and validating before creating/updating? But need to remember every time.

**Comment by gradinarot** [comment]
is there any news on this?

**Comment by malavshah9** [comment]
Yes, I think Prisma should have some method to go for runtime validation.

**Comment by 2color** [comment]
It's worth noting here that even if Prisma were to provide an interface for runtime model validation, Prisma's type system is not quite as complex as the one a validation library would have. 

Prisma's type system is designed to abstract the various nuances between databases thereby simplifying the development experience and providing rich type safety. So even if it were to expose a model-based validation, changes to the PSL (Prisma schema language) would be necessary to validate things like email. Because for the Prisma schema to remain the source of truth, it would need to be able to represent this kind of information.

**Comment by alimehasin** [comment]
How about configure the PrismaClient when insatiate it and pass joi objects, each object will represent a model, PrismaClient will run the validation and either thrown an error if validation failed or complete the execution if it succeed.

As far as I know PrismaClient also do some basic validation and with joi integration we can get include the result of these validation in the same response, I am not sure but unique value validation might be done here and it is great to have unique value errors here.

**Comment by Manubi** [comment]
I also would like to see best practices regarding validation with prisma. :)

**Comment by maxerbox** [comment]
> It's worth noting here that even if Prisma were to provide an interface for runtime model validation, Prisma's type system is not quite as complex as the one a validation library would have.
> 
> Prisma's type system is designed to abstract the various nuances between databases thereby simplifying the development experience and providing rich type safety. If it were to expose a model-based validation, changes to the PSL (Prisma schema language) would be necessary to validate things like email. Because for the Prisma schema to remain the source of truth, it would need to be able to represent this kind of information.

Could be supercool if we could add metadata for generators to fields

**Comment by unlight** [comment]
+1 for custom metadata. There is an issue https://github.com/prisma/prisma/issues/3102 but no movement there

**Comment by GCastilho** [comment]
The ideas presented in #3102 seems really good, like:

```prisma
directive Email {
  type String
  validate <email regex here>
}

directive Salary {
  type Number
  validate v => v > 0
}
```

Then being used like this:

```prisma
model User {
  id    Int @id
  email String @unique @email
  salary Number @salary
}
```

An implicit conversion, say, a numeric string to a number (like mongoose does) would also be nice to prevent from manually having to cast the values or being too strict about types that JS considers "interchangeable" in some cases

**Comment by mmahalwy** [comment]
@GCastilho that's a great solution. 

Another option is to set these types when initiating `PrismaClient` ?

**Comment by GCastilho** [comment]
@mmahalwy you mean like passing this validators when instantiating the `PrismaClient`? But then the prisma schema won't be the "single source of truth" anymore

**Comment by grimen** [comment]
As most of you know there is a widely powerful, standardized, spread, accepted specification for schema validation which is **JSON Schema**, thus not taking that solution/ecosystem into account for Prisma validations would be a mistake IMO. Since Prisma would never be able to replace data validation on all abstractions of a complete system, one would end up with writing validations and/or validation schema generators in more than one place if Prisma would implement an internal validator DSL, which would be unfortunate for such great potential of ecosystem.

**Comment by orenmizr** [comment]
has anyone tried [https://typegraphql.com/](https://typegraphql.com/) ? thoughts ? 
i have yet taken it for a ride but it seem to have class validation + typescript + prisma integration

**Comment by maxerbox** [comment]
> 
> 
> has anyone tried https://typegraphql.com/ ? thoughts ?
> i have yet taken it for a ride but it seems to have class validation + typescript + prisma integration


I'm using it because it generates Classes instead of Types for model objects (works with https://github.com/clayrisser/nestjs-crud-prisma), but it doesn't add runtime validation. Actually, there is an issue mentionning it https://github.com/MichalLytek/typegraphql-prisma/issues/35

**Comment by unlight** [comment]
In prisma-nestjs-graphql generator implemented feature [custom decorators](https://github.com/unlight/prisma-nestjs-graphql#custom-decorators) which allow add `class-validator` decorators. 
It's designed for using with `@nestjs/graphql`, but people are saying that it fits for their needs using REST.

**Comment by aliksend** [comment]
Consider config option to generate raw types (like now) or zod-schemas (with runtime req/resp validation) or joi-schemas etc.

**Comment by matthewmueller** [comment]
We just wrote some documentation on how to add custom validation with Prisma: https://www.prisma.io/docs/concepts/components/prisma-client/custom-validation

We do plan to implement this feature into the Prisma Client eventually, but this is meant to help you along in the meantime. If you'd like to chat about this topic or you have any questions, feel free [to book some time with me](https://calendly.com/matthewmueller/prisma-extend-client).

**Comment by noor-codes** [comment]
Prisma is an ORM which is responsible for managing database, not validations. 
But If Prisma provides a way to add our own custom validations using `validator.js`, `Yup` or other similar libraries instead of doing it on it's own would be a better idea because validations are complex can add up to the Prisma bundle.

We can live without it but If Prisma can make the experience better. That would be great. No Harm done!

**Comment by gsamal** [comment]
I have an use-case where I am trying work on a dynamic query language for my app. It is going to be used with `findMany` and will accepts options such as `table`, `fields`, `filters` etc. I am working on an utility that will convert a custom schema to the format that Prisma accepts. Taking the simple example of [User](https://www.prisma.io/docs/concepts/components/prisma-schema/data-model) model from Prisma docs, the typescript type should be `UserFindManyArgs` (imported from prisma-client).

But the object I am building is dynamic and I wish to validate the final object with the type based on table name. F.e. if table name is `User`, the final options should be of type `UserFindManyArgs`. Is there any way to do it? I have checked out [io-ts](https://github.com/gcanti/io-ts/blob/master/index.md) but finding it impossible to validate against existing type/interface.

**Comment by gsamal** [comment]
@2color I agree with your original suggested approach, it will be beneficial for a lot of cases. I would also like to have `validate` feature of other payloads such as the one we can feed into `FindMany`.
Ref.
```
const result = await prisma.user.findMany({
  where: {
    email: {
      endsWith: 'prisma.io',
    },
    posts: {
      some: {
        published: true,
      },
    },
  },
  include: {
    posts: {
      where: {
        published: true,
      },
    },
  },
})
```

**Comment by okezieobi** [comment]
> Prisma is an ORM which is responsible for managing database, not validations. 
> But If Prisma provides a way to add our own custom validations using `validator.js`, `Yup` or other similar libraries instead of doing it on it's own would be a better idea because validations are complex can add up to the Prisma bundle.
> 
> We can live without it but If Prisma can make the experience better. That would be great. No Harm done! 

My thoughts exactly, Prisma already makes types based on its models accessable, there are libraries that can build validations based on those types like class-validator or typescript-ts.

Fun fact, I have been ORM hopping for the last few weeks, from Sequelize to TypeORM to here and imho Prisma has the most unique take on what an ORM should be concerned with, I like it the way it is, it's lean and mean, like Rocky right before he knocks out Ivan 😁

**Comment by MentalGear** [comment]
I think the big selling point of @Prisma is that it is a one-stop solutions for many database interaction features.

If validators could be used on the frontend, it would mean less work for devs to keep validation in sync since we can use 1 source of truth, which would make Prisma the more valuable.

I agree with the point that there are great validation libs out there which should be used in Prisma to hit the ground running instead of reinventing the wheel.

It would be amazing if we could just import { validation } and use the same rules for live form validation.

**Comment by MentalGear** [comment]
I think the big selling point of @Prisma is that it is a one-stop solutions for many database interaction features.

If validators could be used on the frontend, it would mean less work for devs to keep validation in sync since we can use 1 source of truth, which would make Prisma the more valuable.

I agree with the point that there are great validation libs out there which should be used in Prisma to hit the ground running instead of reinventing the wheel.

It would be amazing if we could just import { validation } and use the same rules for live form validation.

**Comment by MentalGear** [comment]
Just found out about Generators and community plugins. These might be what we are looking for:
https://www.prisma.io/docs/concepts/components/prisma-schema/generators#community-generators

Especially:
- [prisma-json-schema-generator](https://github.com/valentinpalkovic/prisma-json-schema-generator): @grimen 
- [prisma-joi-generator](https://github.com/omar-dulaimi/prisma-joi-generator): Generate full Joi schemas from your Prisma schema.

Would be great if the guide would have an explicit page on how to convert validation to be used for front end form validation using generate and community plugins. @matthewmueller

**Comment by collinscangarella** [comment]
Here's how apollo is tackling the exact same problem: https://www.apollographql.com/blog/backend/validation/graphql-validation-using-directives/

**Comment by millsp** [comment]
Hey everyone, we just shared a proposal about Prisma Client Extensions. While this isn't going to perform validations via the schema, it will provide a way for you to store custom validation functions on your models (and perhaps create a generic one). Here's how it would work:

```ts
const prisma = new PrismaClient().$extends({
	$model: {
		User: {
			validate(user: object) {
				// ... Use Joi, or Yup here
			},
		},
	}
})

await prisma.user.validate(randomData)
```

I am interested in receiving your feedback and suggestions in the [Prisma Client Extensions proposal](https://github.com/prisma/prisma/issues/15074). Thanks!

**Comment by MentalGear** [comment]
Good to see Prisma is taking this issue seriously!
I'm sure omar-dulaimi, maker of various validation generators, has some thoughts here. https://github.com/omar-dulaimi

**Comment by omar-dulaimi** [comment]
Thanks for the shoutout @MentalGear 

Recently, I have been thinking about doing validations in Prisma like this:

```
generator client {
  provider            = "prisma-client-js"
  validationsProvider = "joi"
  validationsOutput   = "../src/validations"
  validationsMode     = "AUTO_RUN" // or GENERATE_ONLY
}
```

When generating the client, it also generates full crud schemas  of the provider specified above, in the output directory also specified.

Once they get generated, we can then incorporate them inside the suggested model methods way to do validations. It would be as simple as importing and executing them.

Relying on known libraries like Joi, Zod, Yup, etc helps reduce the amount of work required to redo what they have been doing for years. So I see it as a plus for the Prisma team.

### Possible structure  
```
package.json
src => 
      validations =>
                        index.ts
                        User =>
                                index.ts
                                createUser.schema.ts
                        Post =>
                                index.ts
                                updatePost.schema.ts
```

### AUTO_RUN Mode


When this mode is selected all generated schemas will be run automatically before actions
with the same name get executed. This of course means that the user does not really need
or depend on the result of the validations; they just need to make sure data are correct before getting inserted/queried. 

Which helps abstracting the entire validation layer away from the user and let them focus on the business logic instead.



### GENERATE_ONLY Mode

This mode does exactly what it means; generate the validations schemas and stop there.
It will be then the users' responsibility to manually run them:

```
// does not throw
const { result, error } = await prisma.user.validate(randomData);
```

Because this user either wants access to validation results or to customize/add more schemas.

This can be done using the validate method described in the proposal:



```ts
import { UserCreateOneSchema } from "./validations"

const prisma = new PrismaClient().$extends({
  $model: {
    User: {
      async validate(user: object, opName: string) {
        try {
          if (opName === "createSuperCustomOne") {
            const result = await UserCreateOneSchema.validateAsync(user);
            return { result, error: null };
          }
        } catch (error) {
          return { result: null, error };
        }
      },
    },
  }
})

const { result, error } = await prisma.user.validate(randomData)
```

The example above should not completely replace the validate method defined by Prisma under the hood; it's just meant to add to whatever already there. So, default actions won't be affected unless overriding them was intentional. Maybe it would make more sense to rename the method to `attach`, `add` or something more appropriate. Or even push to an array.

Also, the Prisma client or the generator in charge of generating the validation schemas, should not delete any custom schemas made by the user.

Would love to hear your thoughts on this.

**Comment by Yukimotochern** [comment]
Would be better if the prisma client can export both type and jsonschema (something like typebox), such that the user can choose their own library, like ajv or so, for validation.

**Comment by MentalGear** [comment]
Any updates / feedback on validation, @millsp ?

**Comment by miketromba** [comment]
Here are a bunch of runtime validation API ideas:

### Model-level validation
```js
prisma.User.$useValidation({
    name: (value) => {
        // Custom validation logic here (Use whatever... zod, joi, etc.)
        // Throw with error message if invalid or return if valid
    }
    // ...optionally define other fields
})
```

### Support async validation
```js
prisma.User.$useValidation({
    name: async (value) => {
        // Perform async queries before making a decision
    }
})
```

### Native integrations with popular validation libraries for even better DX
For inspiration, see how tRPC supports this feature https://trpc.io/docs/v10/procedures#with-zod
```js
// E.g. with zod
import { z } from 'zod'

prisma.User.$useValidation({
    name: z.string().min(2).max(100)
})

// Alternative
prisma.User.$useValidation(
    z.object({
        name: z.string().min(2).max(100)
    })
)
```

### Bonus (prob overkill): support user-injected context object
```js
// Example use case: you might inject a request object
// ...or maybe your validation of one field depends on other the value of another field...
prisma.User.$useValidation({
    name: async (value, ctx) => {
        // Same as above
    }
})

// Then when calling .delete, optionally inject context
await prisma.User.delete({
    // ...specify delete inputs
    context: { req } // E.g. include context object
})
```

### Schema-level validation declarations
An even better solution might be to support schema-level validation declarations like so:
```prisma
model User {
  id      Int      @id @default(autoincrement())
  email   String   @unique @validate('isEmail') @validate('min', 2) @validate('max', 100) // Declare runtime validations
  email2  String   @unique @validate('isEmail', ['min', 2], ['max', 100]) // Or combined syntax
  name    String?

  // Support compound validations ??
  @@validate('emailsHaveSameDomain', [email, email2])
}
```
This way, you get closer to having a single source of truth.

Then, when instantiating the client, the user must provide these definitions or the PrismaClient constructor will throw
```js
const prisma = new PrismaClient({
    validations: {
        isEmail: value => {
            // Throw with error message if invalid
        },
        // OR with first-class support for popular validation libraries (e.g. zod)
        isEmail: z.string().email()
    }
})

// Alternate registration syntax
PrismaClient.useValidator('isEmail', () => {})
```
In other words, allow runtime validation declarations to exist on the schema, and leave implementation details to the end user at the time of client instantiation.

**Comment by miketromba** [comment]
> The ideas presented in #3102 seems really good, like:
> 
> ```prisma
> directive Email {
>   type String
>   validate <email regex here>
> }
> 
> directive Salary {
>   type Number
>   validate v => v > 0
> }
> ```
> 
> Then being used like this:
> 
> ```prisma
> model User {
>   id    Int @id
>   email String @unique @email
>   salary Number @salary
> }
> ```
> 
> An implicit conversion, say, a numeric string to a number (like mongoose does) would also be nice to prevent from manually having to cast the values or being too strict about types that JS considers "interchangeable" in some cases

I like this concept a lot; it would be even better if we could define the implementation of those directives on the client at runtime, similar to my previous comment... This allows users to leverage 3rd party libraries for validation logic instead of rolling their own.
E.g:
```js
const prisma = new PrismaClient({
    validations: {
        email: value => {
            // Throw with error message if invalid
        },
        // OR with first-class support for popular validation libraries (e.g. zod)
        isEmail: z.string().email()
    }
})

// Or maybe
PrismaClient.registerValidator('email', () => {})
```

It would also unlock the ability for open source prisma-validation libraries to exist...
E.g.
```js
import prismaValidations from 'some-community-prisma-validations-plugin'

const prisma = new PrismaClient({
    validations: prismaValidations
})
```

**Comment by tmtron** [comment]
> Bonus (prob overkill): support user-injected context object

IMHO this is not a bonus, but very important for **i18n** - e.g. when you have a multilingual application you must somehow pass information about the current user: i.e. the language, date-time format, number format (decimal separator, etc.).  
e.g. in [class-validator #169](https://github.com/typestack/class-validator/issues/169#issuecomment-616512257) there is still no simple way to translate validation messages

**Comment by shawnmuzick** [comment]
The problem I've run into mostly, is that I have to manually redefine classes/objects. I can write my own validation checks, but I don't think it makes sense to manually define a class or object to validate against. If the Prisma types that are exposed were classes or objects instead of interfaces, I'd be able to make comparisons against it. As it stands now, I'd have to manually create an object that mimicks the model/interface, and update it whenever the schema changes, thus losing the "single source of truth".

**Comment by Marcjazz** [comment]
> ## Problem
> Before saving data to the database, it's common to perform validation.
> 
> When using Prisma to save user data (in an API), Prisma will perform runtime validation before saving to the database and throw an error if the user data doesn't match the type.
> 
> Even with TypeScript, these checks cannot happen at compile time because the user data is not known.
> 
> Since there are situations where you want to validate the data as a separate step to saving to the database, it'd be nice if Prisma exposed the same validation functionality as a standalone method on models.
> 
> ## Suggested solution
> Provide a validate function for models which is based on the Prisma schema:
> 
> ```prisma
> model User {
>   id        Int     @default(autoincrement()) @id
>   email     String  @unique
>   firstName String?
>   lastName  String?
>   social    Json?
>   isAdmin   Boolean @default(false)
> }
> ```
> 
> API ideas:
> 
> ```ts
> const { isValid, error } = prisma.user.validate(user)
> ```
> 
> The idea is that If the input is valid, then the error will be undefined. If the input is invalid, `error` is assigned an Error object providing more information.
> 
> OR
> 
> ```ts
> const user = { firstName: 'Laura' }
> try {
>   const validatedUser = await prisma.user.validate(user)
> } catch (error) {
>   // handle error
> }
> ```
> 
> ## Additional context
> For inspiration, [Joi's validation errors](https://joi.dev/api/?v=17.2.1#validationerror) might be useful (albeit more extensive as Joi has richer validation rules). 

`class-validator` and `class-transformer` packages combine together will eventually solve any runtime type checked. Just in case you are using `NestJS` framework check https://docs.nestjs.com/openapi/mapped-types for more complexes types.

**Comment by Manubi** [comment]
Can we please make it in a way that I write the validation once and I can use it on the DB, the resolvers, and in the frontend? Meaning one single source of truth that can and should be used for validation everywhere. Thank you!

**Comment by Marcjazz** [comment]
> Can we please make it in a way that I write the validation once and I can use it on the DB, the resolvers, and in the frontend? Meaning one single source of truth that can and should be used for validation everywhere. Thank you!

Actually in a mono repo workspace you can write a single library for your type validation and use it on all apps found within the workspace

**Comment by yosietserga** [comment]
I think you could use this from the documentation https://www.prisma.io/docs/concepts/components/prisma-client/client-extensions/model#add-a-custom-method-to-all-models-in-your-schema

**Comment by olivierwilkinson** [comment]
Hi 👋 ,

I've just released a middleware called [prisma-validation-middleware](https://github.com/olivierwilkinson/prisma-validation-middleware) that validates data when creating or updating records, including when doing so through a relation. For example it's possible to validate the `Comment` data used in the `connectOrCreate` below:

```typescript
await client.post.update({
  where: { id: 1 },
  data: {
    comments: {
      connectOrCreate: {
        where: { id: 2 },
        create: [comment data you want to validate]
      },
    },
  },
});
```

It is built to support a javascript validation function and supports using Zod, Superstruct or any 3rd party validation library. There are some examples for [Zod](https://github.com/olivierwilkinson/prisma-validation-middleware#usage-with-zod) and [Superstruct](https://github.com/olivierwilkinson/prisma-validation-middleware#usage-with-superstruct) in the readme.

While this does mean that the Prisma schema is not the only source of truth, it does enable global validation of data while we wait for the `$nestedOperations` API to land in client extensions.

I hope people find it useful! Please don't hesitate to raise any issues / feature requests. 🙏 

p.s. If you are curious how it is handling nested write operations using middleware check out [prisma-nested-middleware](https://github.com/olivierwilkinson/prisma-nested-middleware) which this is built upon 😄

**Comment by kavyantic** [comment]
What's the status of this request. Has it been implemented?

**Comment by dedupely** [comment]
This has been open for 4 years. Any progress here? If not, I'll just create my spinup using zod and extensions. Seems like a really simple thing to have. I mean, if you're going to be coolest ORM in TS why not go that extra 10% and just get there.

What I would like just *.schema level validations that are baked into the model language. Also, having access to those validations so I can re-use them in let's say specific form validations would be great. Makes everything super maintainable. 

Right now I'd have to create an extension and handle over 40 tables and create a structure for storing all of my zod schemas, make sure the correct client is properly imported and extended. It's just not a nice solution and it's also a lot of extra work for something that apparently has a market since 2020.

**Comment by voycey** [comment]
The ship has long sailed on this IMO - people who need it have simply moved onto ORMs that support it like MikroORM that have put QE as a first class feature. Its a shame because the Node ORM ecosystem is very limited in terms of full featured ORMs that are continually maintained

**Comment by Nazaire** [comment]
My approach was creating [zod](https://www.npmjs.com/package/zod) schemas and using that for validation. It works just fine passing a zod type into a prisma type if they're the same shape. For enums you can use zod's nativeEnum type and pass in the generated prisma enum type, works perfect. 

Quite often you want to have a separate schema layer for the the API anyway so this wasn't a major issue for me in the end. Having a separate schema definition ensures you're not accidentally exposing db fields, which might happen if the schema was being generated from the prisma types.

Prisma and Zod, both essential tools for me now.

**Comment by MentalGear** [comment]
do you have a more detailed description of this approach may be somewhere?On 20 Aug 2024, at 05:24, Dylan Kay ***@***.***> wrote:﻿
My approach was creating zod schemas and using that for validation. It works just fine passing a zod type into a prisma type if they're the same shape. For enums you can use zod's nativeEnum type and pass in the generated prisma enum type, works perfect.
Quite often you want to have a separate schema layer for the the API anyway so this wasn't a major issue for me in the end. Having a separate schema definition ensures you're not accidentally exposing db fields, which might happen if the schema was being generated from the prisma types.
Prisma and Zod, both essential tools for me now.

**Comment by Digital39999** [comment]
> My approach was creating [zod](https://www.npmjs.com/package/zod) schemas and using that for validation. It works just fine passing a zod type into a prisma type if they're the same shape. For enums you can use zod's nativeEnum type and pass in the generated prisma enum type, works perfect.
> 
> Quite often you want to have a separate schema layer for the the API anyway so this wasn't a major issue for me in the end. Having a separate schema definition ensures you're not accidentally exposing db fields, which might happen if the schema was being generated from the prisma types.
> 
> Prisma and Zod, both essential tools for me now.
> 
> HOWEVER it would still be great to know ahead of time in some cases whether the objects we're passing in to Prisma are going to fail validation! For example, handling bulk inserts and identifying the records that fail validation.

It would also be really great if prisma somehow leverages zod internally, because keeping 2 schemas up to date is a little hard on very big projects. All tools/generators that have tried that are now old and outdated which is sad.

**Comment by rnenjoy** [comment]
Still not implemented?

**Comment by BOSSgua** [comment]
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/1
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/2
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/3
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/4
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/5
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/6
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/7
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/8
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/9
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/10
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/11
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/12
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/13
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/14
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/15
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/16
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/17
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/18
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/19
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/20
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/21
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/22
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/23
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/24
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/25
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/26
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/27
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/28
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/29
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/30
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/31
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/32
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/33
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/34
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/35
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/36
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/37
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/38
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/39
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/40
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/41
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/42
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/43
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/44
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/45
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/46
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/47
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/48
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/49
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/50
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/51
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/52
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/53
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/54
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/55
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/56
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/57
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/58
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/59
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/60
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/61
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/62
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/63
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/64
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/65
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/66
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/67
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/68
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/69
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/70
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/71
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/72
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/73
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/74
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/75
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/76
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/77
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/78
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/79
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/80
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/81
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/82
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/83
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/84
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/85
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/86
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/87
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/88
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/89
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/90
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/91
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/92
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/93
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/94
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/95
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/96
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/97
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/98
https://github.com/linbojuezhiming/linbojuezhiming.github.io/issues/99
